src/Microsoft.AspNetCore.Razor/CodeGenerators/Visitors/CSharpTagHelperFieldDeclarationVisitor.cs

```diff
@@ -95,32 +96,39 @@ protected override void Visit(TagHelperChunk chunk)
 
             if (!Context.Host.DesignTimeMode)
             {
-                PreAllocateUnboundTagHelperAttributes(chunk);
+                PreAllocateTagHelperAttributes(chunk);
             }
 
             // We need to dive deeper to ensure we pick up any nested tag helpers.
             Accept(chunk.Children);
         }
 
-        private void PreAllocateUnboundTagHelperAttributes(TagHelperChunk chunk)
+        private void PreAllocateTagHelperAttributes(TagHelperChunk chunk)
         {
             var boundAttributes = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
             for (var i = 0; i < chunk.Attributes.Count; i++)
             {
                 var attribute = chunk.Attributes[i];
-                var hasAssociatedDescriptors = chunk.Descriptors.Any(descriptor =>
-                    descriptor.Attributes.Any(attributeDescriptor => attributeDescriptor.IsNameMatch(attribute.Name)));
+                var associatedAttributeDescriptors = chunk.Descriptors.SelectMany(descriptor => descriptor.Attributes)
+                    .Where(attributeDescriptor => attributeDescriptor.IsNameMatch(attribute.Name));
 
                 // If there's no descriptors associated or we're hitting a bound attribute a second time.
-                if (!hasAssociatedDescriptors || !boundAttributes.Add(attribute.Name))
+                var isUnBoundAttribute = !associatedAttributeDescriptors.Any() || !boundAttributes.Add(attribute.Name);
```

Comment:
Given the use of this `bool` later, probably better to move `!boundAttributes.Add(attribute.Name)` into the `if` and have this variable just mean the attribute isn't bound.

If on the other hand `TagHelperAttributeKey` needs this meaning, suggest renaming the variable to describe the actual condition e.g. `isUnboundOrRepeatedAttribute`.